runtime/shared_heap.c

```diff
@@ -789,6 +807,482 @@ void caml_verify_heap(caml_domain_state *domain) {
+    Programming Language LISP). At a high level the algorithm works as a series
+    of parallel (using all running domains) phases separated by global barriers:
+
+    1. For each size class
+      a. Compute the number of live blocks in partially filled pools
+      b. Keep enough pools to fully contain the number of live blocks and
+         set the rest to be evacuated
+      c. For each live block in each pool in the evacuation list,
+         allocate and copy into a non-evacuating pool.
+    2. Proceed through the roots and the heap, updating pointers to evacuated
+       blocks to point to the new location of the block. Update finalisers and
+       ephemerons too.
+    3. Go through pools evacuated and release them. Finally free all but
+        one pool in the freelist.
+    4. One domain needs to release the pools in the freelist back to the OS.
+
+    The algorithm requires one full pass through the whole heap (pools and large
+    allocations) to rewrite pointers, as well as two passes through the
+    partially-occupied pools in the heap to compute the number of live blocks
+    and evacuate them.
+  */
+
+  /* First phase. Here we compute the number of live blocks in partially
+  filled pools, determine pools to be evacuated and then evacuate from them.
+  For the first phase we need not consider full pools, they
+  cannot be evacuated to or from. */
+  caml_global_barrier();
+  CAML_EV_BEGIN(EV_COMPACT_EVACUATE);
+
+  struct caml_heap_state* heap = Caml_state->shared_heap;
+
+  #ifdef DEBUG
+  /* Check preconditions for the heap: */
+  for (int sz_class = 1; sz_class < NUM_SIZECLASSES; sz_class++) {
+    /* No sweeping has happened yet */
+    CAMLassert(heap->avail_pools[sz_class] == NULL);
+    CAMLassert(heap->full_pools[sz_class] == NULL);
+    CAMLassert(heap->swept_large == NULL);
+    /* No pools waiting for adoption */
+    if (participants[0] == Caml_state) {
+      CAMLassert(pool_freelist.global_avail_pools[sz_class] == NULL);
+      CAMLassert(pool_freelist.global_full_pools[sz_class] == NULL);
+    }
+    /* The minor heap is empty */
+    CAMLassert(Caml_state->young_ptr == Caml_state->young_end);
+    /* The mark stack is empty */
+    CAMLassert(caml_mark_stack_is_empty());
+  }
+  #endif
+
+  /* All evacuated pools (of every size class) */
+  pool *evacuated_pools = NULL;
+
+  for (int sz_class = 1; sz_class < NUM_SIZECLASSES; sz_class++) {
+    /* We only care about moving things in pools that aren't full (we cannot
+    evacuate to or from a full pool) */
+    pool* cur_pool = heap->unswept_avail_pools[sz_class];
+
+    if (!cur_pool) {
+      /* No partially filled pools for this size, nothing to do */
+      continue;
+    }
+
+    /* count the number of pools */
+    int num_pools = 0;
+
+    while (cur_pool) {
+      num_pools++;
+      cur_pool = cur_pool->next;
+    }
+
+    struct compact_pool_stat* pool_stats = caml_stat_alloc_noexc(
```

Comment:
This allocation can fail.

It seems that the routine can nevertheless work correctly if `pool_stats` is `NULL`, it will simply be more aggressive since the `free_blocks` and `remaining_live_blocks` counters won't be computed, so it should be relatively easy to cope with allocation failure here.

Reply:
Good point. I'll get this fixed.

Reply:
This is done in https://github.com/ocaml/ocaml/pull/12193/commits/354f0116d48dd0d33333882ad738c6ef66a120b8